Add dynamic primitive clipping example

pumexx opened the pull request:
## Description

This example demonstrates how to use primitive clipping through dynamic pipeline state. 

Dynamic properties required to implement this example are defined in VK_EXT_extended_dynamic_state3 extension and in chapter 27.4 of the Vulkan specification ( "Primitive Clipping" ).

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/README.adoc)
- [x] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [x] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

pumexx committed:
Add dynamic primitive clipping example

pumexx commented:
There's one error present in this example: when I switch primitive clipping off in a GUI and subsequently code calls `vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE)`, the primitive clipping is still enabled. 
At first I thought that it may be an error in nVidia driver. But AMD driver shows the same error. I went through Vulkan specification checking other possible paths that may turn the primitive clipping on and did not find any clue on why this error happens.

I modified `shader_object` example to use `vkCmdSetDepthClipEnableEXT` and got the same error - calls to this command are ignored.

I also consulted my friend from Vulkan CTS group and we established that `vkCmdSetDepthClipEnableEXT` command is not tested at all ( to be precise: function is called once in tests, but `gl_ClipDistance` builtin is not used in the shaders used by the test ). 

For me it looks like `vkCmdSetDepthClipEnableEXT` command was not implemented in both NVidia and AMD drivers at all.
Both drivers just use the results of 'gl_ClipDistance' calculation without checking additional condition set by `vkCmdSetDepthClipEnableEXT`.
It may be because primitive clipping is a very rare animal.

I plan to create an issue in Vulkan CTS project, so that `vkCmdSetDepthClipEnableEXT` has proper test coverage.

SaschaWillems commented:
Thank you very much for this sample. New samples are always very much appreciated :)

Right now CI fails due to outdated copyrights and problems with clang format. Once those are fixed, we'll start reviewing it. If you need any assistance, feel free to ping.

pumexx committed:
Fix issues found by continous integration

SaschaWillems commented:
> There's one error present in this example: when I switch primitive clipping off in a GUI and subsequently code calls `vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE)`, the primitive clipping is still enabled.
> At first I thought that it may be an error in nVidia driver. But AMD driver shows the same error. I went through Vulkan specification checking other possible paths that may turn the primitive clipping on and did not find any clue on why this error happens.

This seems to work fine for me. If I uncheck "use clipping", it looks like this on my 4070 with recent Vulkan developer drivers:

![image](https://github.com/KhronosGroup/Vulkan-Samples/assets/10283127/cd414567-66f5-4f9f-8979-516798a97607)

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/README.adoc:
```diff
@@ -0,0 +1,103 @@
+TIP: The source for this sample can be found in the https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/dynamic_depth_clipping[Khronos Vulkan samples github repository].
+endif::[]
+
+== Overview
```
Can you add a screenshot of the sample in here? See e.g. the Dynamic line rasterization sample.

pumexx replied:
Added a screenshot.

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/README.adoc:
```diff
@@ -0,0 +1,103 @@
+-
+////
+
+== Dynamic primitive clipping
```
Header level should be one level higher, so instead of '==' use '='. Otherwise our docs site build process will output a warning due to a mismatch in header levels.

pumexx replied:
Done.

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.h:
```diff
@@ -0,0 +1,104 @@
+	{
+		VkDescriptorSet descriptor_positive;
+		VkDescriptorSet descriptor_negative;
+	} descriptor_sets;
```
These should be initialized to `VK_NULL_HANDLE`. MSVC will output a warning if not.

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.h:
```diff
@@ -0,0 +1,104 @@
+	struct
+	{
+		VkDescriptorSetLayout models;
+	} descriptor_set_layouts;
```
This should be initialized to `VK_NULL_HANDLE`. MSVC will output a warning if not.

pumexx replied:
Done.

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.h:
```diff
@@ -0,0 +1,104 @@
+	struct
+	{
+		VkPipelineLayout models;
+	} pipeline_layouts;
```
This should be initialized to `VK_NULL_HANDLE`. MSVC will output a warning if not.

pumexx replied:
Done.

SaschaWillems commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.h:
```diff
@@ -0,0 +1,104 @@
+		bool    useClipping;
+		bool    drawObject[2];
+		int32_t visualization;
+	} params;
```
This should be (zero) initialized . MSVC will output a warning if not.

pumexx replied:
Done. Also I removed code initializating the same values in DynamicPrimitiveClipping::prepare()

SaschaWillems reviewed (changes requested):
Looks good to me so far.

I only have a few very minor remarks added as comments.

Thanks for adding this new sample :)

pumexx committed:
Add screenshot to README.adoc. Fix object initialization issues.

pumexx commented:
> 
> This seems to work fine for me. If I uncheck "use clipping", it looks like this on my 4070 with recent Vulkan developer drivers:
> 

In that case we should see the whole Utah teapot as "blue". But it looks like a call to `vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE)` is ignored.

I could hide "use clipping" checkbox in order to disguise the error (at least temporarily), but I think this sample may be a good place for field testing by driver developers.

SaschaWillems commented:
This oddly still fails with CI checks. @tomadamatkinson: Any idea? It seems to fail at files not even part of this PR

pumexx commented:
> This oddly still fails with CI checks. @tomadamatkinson: Any idea? It seems to fail at files not even part of this PR

That's why I didn't fix it yet. I am afraid of avalanche of another wrong copyright checks.

pdaniell-nv commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.cpp:
```diff
@@ -0,0 +1,384 @@
+
+		// Turn the primitive clipping off, so that it does not interfere with further rendering ( GUI rendering comes next )
+		// It may interfere, because using dynamic states turns Vulkan pipelines into state machines similar to OpenGL.
+		vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE);
```
This line shouldn't be necessary as the "draw_ui" has it's own pipeline which sets this state statically from the pipeline.

pumexx replied:
Line removed.

pdaniell-nv commented:
Reading the comments in this PR it makes me wonder if ClipDistance is being confused with depth clipping, which are different steps in the primitive clipping part of the pipeline. User defined clip distances are enabled/disable by the presence of ClipDistance decorated outputs in the shader, and are independent to the depth clamp and depth clip pipeline states. In other words, setting vkCmdSetDepthClipEnableEXT(VK_FALSE) or vkCmdSetDepthClipEnableEXT(VK_TRUE) won't have any affect on any ClipDistance specified in the shader and their operation.

gary-sweet commented:
I'm surprised this even ran for me since we don't support some of the features being used here. I suspect you may be missing some checks. See https://vulkan.gpuinfo.org/displayreport.php?id=27052 for the full set of device support on the board I'm using.

I also see some validation errors which are likely related:
`[error] [framework/core/instance.cpp:50] -1876993556 - VUID-VkDeviceCreateInfo-pNext-pNext: Validation Error: [ VUID-VkDeviceCreateInfo-pNext-pNext ] Object 0: handle = 0x2a78ced0, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x901f59ec | vkCreateDevice(): pCreateInfo->pNext includes a pointer to a VkPhysicalDeviceDepthClipEnableFeaturesEXT, but when creating VkDevice, the parent extension (VK_EXT_depth_clip_enable) was not included in ppEnabledExtensionNames.`

`[error] [framework/core/instance.cpp:50] 115483881 - VUID-VkShaderModuleCreateInfo-pCode-08740: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08740 ] | MessageID = 0x6e224e9 | vkCreateGraphicsPipelines(): pCreateInfos[0].pStages[0] SPIR-V Capability ClipDistance was declared, but one of the following requirements is required (VkPhysicalDeviceFeatures::shaderClipDistance). The Vulkan spec states: If pCode is a pointer to SPIR-V code, and pCode declares any of the capabilities listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08740)`

pumexx committed:
Decoupled primitive clipping from depth clipping.

Added primitive clipping using half-spaces.
Updated documentation to contain latest changes.

pumexx committed:
Minor documentation fixes.

pumexx committed:
Modified README.adoc title

pumexx commented:
> Reading the comments in this PR it makes me wonder if ClipDistance is being confused with depth clipping, which are different steps in the primitive clipping part of the pipeline. 

You are right @pdaniell-nv . I mixed up these two steps. 

I was looking for some external switch that may cause `gl_ClipDistance` computations to be ignored like we can ignore depth values computed in a shader by turning off depth writing and depth testing in a graphics pipeline. `vkCmdSetDepthClipEnableEXT` looked like a best match, especially because it's defined in a Vulkan specification in the same chapter right after primitive clipping which uses `gl_ClipDistance`.

I decoupled these two features in my sample and now it presents both of them. I also described both of these features in README.adoc and I specifically emphasized that these two are distinct features. I hope this helps to solve potential user confusion.

@gary-sweet :

> I'm surprised this even ran for me since we don't support some of the features being used here. I suspect you may be missing some checks.

I added additional check: if `shaderClipDistance` feature is not present then application is terminated.

@SaschaWillems: I fixed these two strange copyright checks in a code that I did not touch and now I see that there is some conflict to resolve. Great...

SaschaWillems commented:
@tomadamatkinson: can you take a look at this? Having to change copyrights on files not actually touched by a PR is something we shouldn't require people to do.

pumexx committed:
Merge branch 'main' into dynamic_primitive_clipping

pumexx committed:
Fixed visualisation names in GUI.

SaschaWillems reviewed (dismissed):
LGTM and works fine for me on Win11 and an RTX4070

asuessenbach commented on samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.cpp:
```diff
@@ -0,0 +1,396 @@
+			vkCmdSetDepthClipEnableEXT(cmd, VK_TRUE);
+		else
+			vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE);
+
```
Maybe simply
`vkCmdSetDepthClipEnableEXT(cmd, params.useDepthClipping)`

SaschaWillems replied:
Yes, that's valid. `VK_EXT_depth_clip_enable` is an extension to enable a D3D feature. Not sure if we want to use or demonstrate that in a sample.

pumexx replied:
Valid usage section for `vkCmdSetDepthClipEnableEXT()` command states that:

> VUID-vkCmdSetDepthClipEnableEXT-depthClipEnable-07451
 The `depthClipEnable` feature must be enabled

`depthClipEnable` feature is defined inside `VkPhysicalDeviceDepthClipEnableFeaturesEXT` structure which is part of `VK_EXT_depth_clip_enable` extension.

At first I didn't add this extension to the list, but then @gary-sweet  pointed out that it emits a validation error:

> [error] [framework/core/instance.cpp:50] -1876993556 - VUID-VkDeviceCreateInfo-pNext-pNext: Validation Error: [ VUID-VkDeviceCreateInfo-pNext-pNext ] Object 0: handle = 0x2a78ced0, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x901f59ec | vkCreateDevice(): pCreateInfo->pNext includes a pointer to a VkPhysicalDeviceDepthClipEnableFeaturesEXT, but when creating VkDevice, the parent extension (VK_EXT_depth_clip_enable) was not included in ppEnabledExtensionNames.

pumexx replied:
@asuessenbach 

Should I use
vkCmdSetDepthClipEnableEXT(cmd, params.useDepthClipping);
or maybe 
vkCmdSetDepthClipEnableEXT(cmd, params.useDepthClipping?VK_TRUE:VK_FALSE);

I know that your proposition will work, but still `bool` and `VkBool32` are different data types.

asuessenbach replied:
I think, both are fine.
And to prevent implicit conversion you could as well say
`vkCmdSetDepthClipEnableEXT(cmd, !!params.useDepthClipping);`

pumexx replied:
I decided to use your initial proposition.

asuessenbach reviewed (commented):
This is, what VVL tells me:
`Attempting to enable extension VK_EXT_depth_clip_enable, but this extension is intended to support D3D emulation layers, and applications ported from D3D, by adding functionality specific to D3D and it is strongly recommended that it be otherwise avoided.`
Is that valid? Does anybody know, why it should not be used?

pumexx committed:
Merge branch 'main' into dynamic_primitive_clipping

pumexx committed:
Fix minor boolean issue in vkCmdSetDepthClipEnableEXT()

Mateusz-Kowalewski-Mobica committed:
Adjust sample to changes from PR#910

Mateusz-Kowalewski-Mobica committed:
Merge branch 'main' into dynamic_primitive_clipping

Mateusz-Kowalewski-Mobica commented:
@SaschaWillems @asuessenbach Can I request for your reviews for current state of this sample?
I merge it with main after PR#910 submission, so it is prepared for new approach.
I takeover responsibility of it from Paweł.

SaschaWillems reviewed (dismissed):
Works fine for me on Windows 11 with an NVIDIA RTX 4070.

This is a really nice example. Thank you very much 👍🏻

Mateusz-Kowalewski-Mobica committed:
Merge branch 'main' into dynamic_primitive_clipping

asuessenbach commented on samples/extensions/README.adoc:
```diff
@@ -262,8 +262,14 @@ Demonstrate how to use the blending related functions available in the VK_EXT_ex
+*Extension*: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_shader_non_semantic_info.html[`VK_KHR_shader_non_semantic_info`]
+
+Demonstrates how to use https://en.wikipedia.org/wiki/Printf[Printf] statements in a shader to output per-invocation values. This can help find issues with shaders in combination with graphics debugging tools.
+
```
Isn't this text part of a different PR?

Mateusz-Kowalewski-Mobica replied:
Yes, but this is weird. As you see I didn't change any letter even line is the same. After that line I added new line with dynamic primitive clipping. To be more elegant I can fix that to doesn't have that line in git diff.

marty-johnson59 commented:
Merging - 3 approvals
